### PR TITLE
Pebble based chunk data packs storage

### PR DESCRIPTION
**Pull request by zhangchiqing**
Close https://github.com/onflow/flow-go/issues/6040

This PR replace the badger based chunk data pack with pebble based chunk data pack storage.

Benchmarking with testnet data shows a 43% reducing in disk usage.

### CODE REVIEW

File: storage/pebble/operations/chunk_data_pack.go [hunk 1]
```diff
@@ -0,0 +1,25 @@
+package operations
+
+import (
+	"github.com/cockroachdb/pebble"
+
+	"github.com/onflow/flow-go/model/flow"
+	"github.com/onflow/flow-go/storage"
+)
+
+func InsertChunkDataPack(sc *storage.StoredChunkDataPack) func(w pebble.Writer) error {
```
Comment (peterargue) [inline storage/pebble/operations/chunk_data_pack.go:10]: please add godocs including the error cases for these

File: storage/pebble/open.go [hunk 2]
```diff
@@ -47,6 +49,18 @@ func OpenRegisterPebbleDB(dir string) (*pebble.DB, error) {
 	return db, nil
 }
 
+func OpenDefaultPebbleDB(dir string) (*pebble.DB, error) {
```
Comment (peterargue) [inline storage/pebble/open.go:52]: please add godocs

File: storage/pebble/open.go [hunk 3]
```diff
@@ -35,7 +37,7 @@ func NewBootstrappedRegistersWithPath(dir string) (*Registers, *pebble.DB, error
 
 // OpenRegisterPebbleDB opens the database
```
Comment (peterargue) [inline storage/pebble/open.go:38]: can you update these comments to explain the difference between this and `OpenDefaultPebbleDB`

File: storage/pebble/value_cache.go [hunk 4]
```diff
@@ -0,0 +1,148 @@
...
+
+func withLimit[K comparable, V any](limit uint) func(*Cache[K, V]) {
+	return func(c *Cache[K, V]) {
+		c.limit = limit
+	}
+}
+
+type storeFunc[K comparable, V any] func(key K, val V) func(pebble.Writer) error
+
+//	func withStore[K comparable, V any](store storeFunc[K, V]) func(*Cache[K, V]) {
```
Comment (peterargue) [inline storage/pebble/value_cache.go:22]: any reason to leave this commented out?

File: storage/pebble/value_cache.go [hunk 5]
```diff
@@ -0,0 +1,148 @@
...
+	"fmt"
+
+	"github.com/cockroachdb/pebble"
+	lru "github.com/hashicorp/golang-lru/v2"
+
+	"github.com/onflow/flow-go/module"
+	"github.com/onflow/flow-go/storage"
+)
+
+func withLimit[K comparable, V any](limit uint) func(*Cache[K, V]) {
```
Comment (peterargue) [inline storage/pebble/value_cache.go:14]: can you add godocs for these methods explaining how they are used

File: storage/pebble/value_cache.go [hunk 6]
```diff
@@ -0,0 +1,148 @@
...
+//			c.store = store
+//		}
+//	}
+func noStore[K comparable, V any](_ K, _ V) func(pebble.Writer) error {
+	return func(pebble.Writer) error {
+		return fmt.Errorf("no store function for cache put available")
+	}
+}
+
+//	func noopStore[K comparable, V any](_ K, _ V) func(pebble.Reader) error {
```
Comment (peterargue) [inline storage/pebble/value_cache.go:33]: same here

File: storage/pebble/value_cache.go [hunk 7]
```diff
@@ -0,0 +1,148 @@
...
+}
+
+func noRetrieve[K comparable, V any](_ K) func(pebble.Reader) (V, error) {
+	return func(pebble.Reader) (V, error) {
+		var nullV V
+		return nullV, fmt.Errorf("no retrieve function for cache get available")
+	}
+}
+
+type Cache[K comparable, V any] struct {
```
Comment (peterargue) [inline storage/pebble/value_cache.go:53]: can you add godocs explaining what this is used for and how it works?

File: storage/pebble/chunk_data_packs.go [hunk 8]
```diff
@@ -0,0 +1,130 @@
...
+	"github.com/cockroachdb/pebble"
+
+	"github.com/onflow/flow-go/model/flow"
+	"github.com/onflow/flow-go/module"
+	"github.com/onflow/flow-go/module/metrics"
+	"github.com/onflow/flow-go/storage"
+	"github.com/onflow/flow-go/storage/pebble/operations"
+)
+
+type ChunkDataPacks struct {
```
Comment (peterargue) [inline storage/pebble/chunk_data_packs.go:15]: please add godocs for exported methods including expected error cases

**Review by peterargue — APPROVED** [review]
looks good. I added a few comments, mostly about godocs

File: storage/pebble/value_cache.go (hunk 4 above)
Comment (zhangchiqing) [reply storage/pebble/value_cache.go:22]: because it's not being used by any code yet, and will be needed when moving other badger modules to pebble.

**Pull request merged**